File: sipivr.plugins/sipivr.speech/src/main/java/ru/sipivr/speech/core/Noun.java
package ru.sipivr.speech.core;

import org.apache.commons.lang.NotImplementedException;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Karpukhin on 14.01.2016.
 */
public class Noun extends LexicalUnit {
    private String value;
    private String tag;

    /// <summary>
/// whether translator will look at the categories when calculating file name
/// </summary>
    private boolean applyCategories;


    public Noun() {
        this(null);
    }

    public Noun(String value)
    {
        this(null, null);
    }

    public Noun(String tag, String value)
    {
        this(tag, value, true);
    }

    public Noun(String tag, String value, boolean applyCategories) {
        this.value = value;
        this.tag = tag;
        this.applyCategories = applyCategories;
    }

    public String getTag() {
        return tag;
    }

    public String getValue() {
        return value;
    }

    public boolean isApplyCategories() {
        return applyCategories;
    }

    @Override
    public LexicalUnit clone() {
        Noun res = new Noun();
        res.cloneFields(this);
        return res;
    }

    @Override
    protected void cloneFields(LexicalUnit src) {
        Noun n = (Noun) src;
        this.tag = n.tag;
        this.value = n.value;
        this.applyCategories = n.applyCategories;
        super.cloneFields(src);
    }

    @Override
    public List<LexicalUnit> getChildren() {
        throw new NotImplementedException();
    }

    @Override
    public List<LexicalUnit> spell() {
        List<LexicalUnit> res = new ArrayList<>();
        res.add(this);
        return res;
    }
}

File: freeswitch/typescript/src/Utils/Speech/Noun.ts
import { LexicalUnit } from "./LexicalUnit";
import { GrammaticalNumber, Case } from "./Enums";

export class Noun extends LexicalUnit {
    Value: string;
    Tag: string;
    /** whether translator will look at the categories when calculating file name */
    ApplyCategories: boolean;

    constructor(value?: string, tag?: string, applyCategories?: boolean) {
        super();
        this.Value = value === undefined ? null : value;
        this.Tag = tag === undefined ? null : tag;
        this.ApplyCategories = applyCategories === undefined ? true : applyCategories;
    }

    clone(): LexicalUnit {
        let res = new Noun();
        res.cloneFields(this);
        return res;
    }

    cloneFields(src: LexicalUnit) {
        let n = src as Noun;
        this.Tag = n.Tag;
        this.Value = n.Value;
        this.ApplyCategories = n.ApplyCategories;
        super.cloneFields(src);
    }

    getChildren(): LexicalUnit[] {
        throw new Error();
    }

    spell(): LexicalUnit[] {
        let res: LexicalUnit[] = [];
        res.push(this);
        return res;
    }

    getFileName(): string {
        let res: string;
        switch (this.Tag) {
            case "Separator": //слово "целых"
                res = "sipivr.number/int_" + this.Case + "_" + this.GrammaticalNumber;
                break;
            case "Minus":
                res = "sipivr.number/minus";
                break;
            case "Point":
                res = "sipivr.number/point";
                break;
            default:
                let parts: string[] = this.Tag !== null ? [this.Tag] : [];

                if (this.ApplyCategories) {
                    parts.push(this.Case);
                    parts.push(this.GrammaticalNumber);
                }

                parts.push(this.Value);
                res = parts.join(this.Separator);
                break;
        }
        return res + ".wav";
    }

    isNumberNoun(): boolean {
        return false;
    }
}